File: backend/app/api/v2/schemas/intake.py
"""Pydantic request/response DTOs for v2 intake API."""

from __future__ import annotations

from datetime import datetime
from typing import Any, Literal, Optional
from uuid import UUID

from pydantic import BaseModel, Field, model_validator


class IntakeFormData(BaseModel):
    role_name: str = Field(..., min_length=1, max_length=200)
    experience_min: int = Field(..., ge=0, le=50)
    # Optional: None means open-ended ("X+ years"). The requisitions table
    # column is nullable and CHECK (max IS NULL OR max >= min) — mirroring how
    # v1 modelled it. NEVER coerce a blank max to 0; that violates the CHECK
    # whenever min > 0 (the 23514 the create-role flow used to surface).
    experience_max: Optional[int] = Field(None, ge=0, le=50)
    location: str = Field(..., min_length=1, max_length=200)
    jd_text: Optional[str] = Field(None, max_length=50_000)


class CreateSessionRequest(BaseModel):
    # Either a fresh-role form OR an existing requisition to complete intake
    # for (ATS-imported / awaiting publish). With requisition_id the service
    # seeds form_data from the requisition row and form_data is ignored.
    form_data: Optional[IntakeFormData] = None
    requisition_id: Optional[UUID] = None
    entry_point: Optional[str] = Field(
        None,
        description=(
            "One of: create_role_btn | intake_tab | supervisor_handoff | "
            "complete_intake_btn"
        ),
    )

    @model_validator(mode="after")
    def _require_form_or_requisition(self) -> "CreateSessionRequest":
        if self.form_data is None and self.requisition_id is None:
            raise ValueError(
                "form_data is required when requisition_id is not provided"
            )
        return self


class CreateSessionResponse(BaseModel):
    session_id: UUID
    requisition_id: UUID


class ProcessStage(BaseModel):
    name: str
    status: str  # pending | running | completed | failed
    output: Optional[dict[str, Any]] = None
    error: Optional[str] = None
    updated_at: Optional[datetime] = None


class VoiceStartRequest(BaseModel):
    sdp: str = Field(..., description="WebRTC offer SDP")
    type: str = Field("offer", description="WebRTC offer type")


class VoiceStartResponse(BaseModel):
    sdp: str = Field(..., description="WebRTC answer SDP")
    type: str = Field("answer", description="WebRTC answer type")


class IntakeSessionResponse(BaseModel):
    id: UUID
    requisition_id: UUID
    user_id: UUID
    organization_id: UUID
    status: str
    active_modality: Optional[str] = None
    entry_point: Optional[str] = None
    form_data: dict[str, Any]
    questions_version: str
    questions_snapshot: list[dict[str, Any]]
    prefilled_answers: Optional[dict[str, Any]] = None
    current_answers: Optional[dict[str, Any]] = None
    turns: list[dict[str, Any]] = Field(default_factory=list)
    process_stages: list[ProcessStage] = Field(default_factory=list)
    process_status: str = "idle"
    process_error: Optional[str] = None
    interview_plan: Optional[dict[str, Any]] = None
    created_at: datetime
    updated_at: datetime


class PublishRequest(BaseModel):
    """Optional edited plan override. If null, server uses DB-stored interview_plan."""
    interview_plan: Optional[dict[str, Any]] = None


class SubmitResponse(BaseModel):
    session_id: UUID
    status: str


class PublishResponse(BaseModel):
    session_id: UUID
    requisition_id: UUID
    redirect_url: str


# ---------------------------------------------------------------------------
# GET /api/v2/intake/sessions — list view (B1.1)
# ---------------------------------------------------------------------------


class SessionListItem(BaseModel):
    """Single row in the conversation-history list, from user_conversation_history view."""

    session_id: UUID
    requisition_id: UUID
    title: str
    display_status: Literal["incomplete", "submitted", "completed"]
    detail_status: str
    last_activity_at: datetime
    active_modality: Optional[Literal["voice", "text"]] = None
    resume_url: str
    role_name: Optional[str] = None
    exp_min: Optional[int] = None
    exp_max: Optional[int] = None
    location: Optional[str] = None

    # Hub enrichment — goal coverage for the "Continue pending" list
    covered: int = 0
    total: int = 9
    stopped_at: Optional[str] = None

    # Hub enrichment — plan stats for the "Edit existing" list
    rounds_count: int = 0
    total_minutes: int = 0
    candidates_count: int = 0


class ListSessionsResponse(BaseModel):
    sessions: list[SessionListItem]


# ---------------------------------------------------------------------------
# B2.1 — PATCH /api/v2/intake/sessions/{id}/answers
# ---------------------------------------------------------------------------

QuestionIdLiteral = Literal[
    "q1_role_overview", "q2_rounds", "q3_focus_areas", "q4_must_haves",
    "q5_nice_to_haves", "q6_cultural_fit", "q7_team_structure",
    "q8_red_flags", "q9_anything_else",
]

AnswerStatusLiteral = Literal[
    "untouched", "needs_probe", "discussed", "validated", "skipped",
]


class PatchAnswerEntry(BaseModel):
    """Per-question patch — at least one of text|status must be provided."""
    text: Optional[str] = Field(None, max_length=20_000)
    status: Optional[AnswerStatusLiteral] = None

    def model_post_init(self, __context) -> None:
        if self.text is None and self.status is None:
            raise ValueError("patch entry must include text or status")


class PatchAnswersRequest(BaseModel):
    patch: dict[QuestionIdLiteral, PatchAnswerEntry]

    @model_validator(mode="after")
    def _patch_not_empty(self) -> "PatchAnswersRequest":
        if not self.patch:
            raise ValueError("patch must not be empty (min 1 entry required)")
        return self


class PatchAnswersResponse(BaseModel):
    applied: list[QuestionIdLiteral]

File: recruiter-app/src/types/intake.ts
// 1:1 mirror of backend/app/api/v2/schemas/intake.py (verified
// against IntakeFormData, AnswerState, ProcessStage, IntakeSessionResponse,
// SessionListItem). Update both files together when the contract changes.

import type { PersonaSnapshot, ScreeningQuestion } from '@/services/screening';

export type IntakeSessionStatus =
  | 'created'
  | 'prefilling'
  | 'ready'
  | 'active'
  | 'submitted'
  | 'published'
  | 'abandoned';

export type IntakeModality = 'voice' | 'text';

export type AnswerStatus = 'untouched' | 'needs_probe' | 'discussed' | 'validated' | 'skipped';

export type Confidence = 'none' | 'low' | 'medium' | 'high';

export type ProcessStageStatus = 'pending' | 'running' | 'completed' | 'failed';

export type ProcessStatus = 'idle' | 'running' | 'failed';

// MUST stay in sync with intake-core/intake_core/questions.py INTAKE_QUESTIONS ids.
export type QuestionId =
  | 'q1_role_overview'
  | 'q2_rounds'
  | 'q3_focus_areas'
  | 'q4_must_haves'
  | 'q5_nice_to_haves'
  | 'q6_cultural_fit'
  | 'q7_team_structure'
  | 'q8_red_flags'
  | 'q9_anything_else';

export interface IntakeFormData {
  role_name: string;
  experience_min: number;
  // null = open-ended ("X+ years"). Mirrors the nullable requisitions column.
  // NEVER send 0 for a blank max — it trips the DB's max>=min CHECK.
  experience_max: number | null;
  location: string;
  jd_text: string | null;
}

export interface AnswerState {
  status: AnswerStatus;
  text: string | null;
  prefilled_text: string | null;
  extraction_confidence: Confidence;
  sources: string[];
  turns_addressed: number[];
  // Set by B2.1 (Phase 2 PATCH /answers); optional here so the type
  // round-trips both shapes.
  manual_edit?: boolean;
  edited_at?: string;
  edited_by?: string;
}

export type CurrentAnswers = Partial<Record<QuestionId, AnswerState>>;

export interface QuestionSnapshot {
  id: QuestionId;
  order: number;
  topic: string;
  default_text: string;
}

// Phase 1 does NOT render conversation turns. Type included so the session
// row shape matches the backend; downstream phases will populate it.
export interface Turn {
  idx: number;
  role: 'user' | 'assistant';
  content: string;
  modality: IntakeModality;
  timestamp: string;
  duration_ms?: number;
  stt_confidence?: number;
  audio_duration_ms?: number;
  tool_calls?: Array<{
    name: 'update_answer' | 'mark_status';
    args: Record<string, unknown>;
  }>;
}

export interface ProcessStage {
  name:
    | 'check_context'
    | 'parse_jd'
    | 'query_cortex'
    | 'synthesize'
    | 'scorecard_rounds'
    | 'scorecard_details';
  status: ProcessStageStatus;
  output: Record<string, unknown> | null;
  error: string | null;
  updated_at: string | null;
}

export type RoundCategory =
  | 'screening'
  | 'coding'
  | 'design'
  | 'behavioral'
  | 'domain'
  | 'culture'
  | 'panel'
  | 'assessment';

export interface Guideline {
  title: string;
  description: string;
}

export interface FeedbackQuestion {
  id: string;
  question_number: number;
  heading: string;
  description: string | null;
}

// Per-round screening config carried in the plan artifact pre-publish. Stored
// in camelCase (reuses the screening service types + components); mapped to the
// snake_case wire shape only at publish (see `roundScreeningToWire`). Backend
// materializes it at publish — see intake_publish_service._materialize_screening.
export interface RoundScreening {
  enabled: boolean;
  voice: string;
  followUpStyle: string;
  validityDays: number;
  questions: ScreeningQuestion[];
  // The derived persona snapshot (dimensions + composed runtime text), stored
  // verbatim from `derivePersonaDraft`. Absent until the recruiter derives one.
  personaSnapshot?: PersonaSnapshot;
}

export interface InterviewPlanRound {
  id: string;
  round_number: number;
  name: string;
  category: RoundCategory;
  duration_minutes: number;
  description: string;
  skills: string[];
  guidelines: Guideline[];
  feedback_questions: FeedbackQuestion[];
  // Optional pre-publish screening config (the editor sets this; publish
  // materializes it). Absent on rounds with no screening configured.
  screening?: RoundScreening;
}

export interface InterviewPlan {
  rounds: InterviewPlanRound[];
}

export interface IntakeSession {
  id: string;
  requisition_id: string;
  user_id: string;
  organization_id: string;
  status: IntakeSessionStatus;
  active_modality: IntakeModality | null;
  entry_point: string | null;
  form_data: IntakeFormData;
  questions_version: string;
  questions_snapshot: QuestionSnapshot[];
  prefilled_answers: CurrentAnswers | null;
  current_answers: CurrentAnswers | null;
  turns: Turn[];
  process_stages: ProcessStage[];
  process_status: ProcessStatus;
  process_error: string | null;
  interview_plan: InterviewPlan | null;
  created_at: string;
  updated_at: string;
}

export interface SessionListItem {
  session_id: string;
  requisition_id: string;
  title: string;
  display_status: 'incomplete' | 'submitted' | 'completed';
  detail_status: IntakeSessionStatus;
  last_activity_at: string;
  active_modality: IntakeModality | null;
  resume_url: string;
  role_name?: string | null;
  exp_min?: number | null;
  exp_max?: number | null;
  location?: string | null;
  // Hub enrichment — goal coverage for the "Continue pending" list
  covered: number;
  total: number;
  stopped_at?: string | null;
  // Hub enrichment — plan stats for the "Edit existing" list
  rounds_count: number;
  total_minutes: number;
  candidates_count: number;
}

export interface ListSessionsResponse {
  sessions: SessionListItem[];
}

export interface CreateSessionResponse {
  session_id: string;
  requisition_id: string;
}

// All stages enumerated for type-safety; Phase 1 ships only the first four
// (intake_form, prefilling, ready_choose_modality, prefill_failed) — the
// rest are defined so flow.ts deriveStage compiles cleanly and downstream
// phases plug in without changing this file.
export type IntakeStage =
  | 'lobby'
  | 'intake_form'
  | 'prefilling'
  | 'ready_choose_modality'
  | 'prefill_failed'
  | 'voice_active'
  | 'voice_reconnect'
  | 'text_active'
  | 'wrapping'
  | 'plan_generating'
  | 'plan_editing'
  | 'scorecard_failed'
  | 'published';

const _STAGES: ReadonlySet<IntakeStage> = new Set<IntakeStage>([
  'lobby',
  'intake_form',
  'prefilling',
  'ready_choose_modality',
  'prefill_failed',
  'voice_active',
  'voice_reconnect',
  'text_active',
  'wrapping',
  'plan_generating',
  'plan_editing',
  'scorecard_failed',
  'published',
]);

export function isIntakeStage(v: unknown): v is IntakeStage {
  return typeof v === 'string' && _STAGES.has(v as IntakeStage);
}

export function isProcessStage(v: unknown): v is ProcessStage {
  if (v === null || typeof v !== 'object') return false;
  const r = v as Record<string, unknown>;
  return (
    typeof r.name === 'string' && typeof r.status === 'string' && 'output' in r && 'error' in r
  );
}

export function isSessionListItem(v: unknown): v is SessionListItem {
  if (v === null || typeof v !== 'object') return false;
  const r = v as Record<string, unknown>;
  return (
    typeof r.session_id === 'string' &&
    typeof r.requisition_id === 'string' &&
    typeof r.title === 'string' &&
    typeof r.display_status === 'string' &&
    typeof r.resume_url === 'string'
  );
}

export type TextStreamEvent =
  | { type: 'text'; chunk: string }
  | { type: 'tool'; name: string; args: Record<string, unknown> }
  | {
      type: 'done';
      text: string;
      stop_reason: string | null;
      user_turn_idx: number;
      assistant_turn_idx: number;
    }
  | { type: 'error'; message: string; code?: string };

// ─── Phase 4: Handoff + Process Till Now ─────────────────────────────────────

export interface ModalityConflictBody {
  detail: string;
  held: 'voice' | 'text';
  requested: 'voice' | 'text';
}

export interface SwitchToTextResponse {
  active_modality: 'text';
  drained: {
    drained: boolean;
    already_idle?: boolean;
    duration_ms?: number;
    forced_cancel?: boolean;
  };
  session: IntakeSession | null;
}

export interface SwitchToVoiceResponse {
  active_modality: 'voice';
  next_step: 'voice_start';
}

export type SwitchModalityResponse = SwitchToTextResponse | SwitchToVoiceResponse;

// VoiceDrainFailedError + ReprocessAlreadyRunningError as runtime classes
// (these need to be exported from types so the api.ts consumer can throw them
// and tests can `instanceof` them).
export class VoiceDrainFailedError extends Error {
  readonly innerError: string | undefined;
  constructor(detail: string, innerError?: string) {
    super(detail);
    this.name = 'VoiceDrainFailedError';
    this.innerError = innerError;
  }
}

export interface ReprocessAcceptedResponse {
  session_id: string;
  process_run_id: string;
}

export class ReprocessAlreadyRunningError extends Error {
  constructor(detail: string) {
    super(detail);
    this.name = 'ReprocessAlreadyRunningError';
  }
}

export interface HeartbeatResponse {
  ok: true;
  last_heartbeat_at: string;
}
